transport.go

```diff
@@ -150,13 +150,31 @@ func (t *transport) handle() {
 	for {
 		select {
 		case req := <-t.requests:
-			// BUG(stevvooe): This is an awful tag allocation procedure.
-			// Replace this with something that let's us allocate tags and
-			// associate data with them, returning to them to a pool when
-			// complete. Such a system would provide a lot of information
-			// about outstanding requests.
-			tags++
-			fcall := newFcall(tags, req.message)
+			var selected Tag
+			if req.message.Type() == Tversion {
+				// Tversion can only use NOTAG.
+				if _, exists := outstanding[NOTAG]; exists {
+					req.err <- errors.New("NOTAG already in use")
+				}
+				selected = NOTAG
+			} else {
+				// The tag pool is depleted if all 65536 tags are taken, or if
+				// 65535 tags are taken and NOTAG is available.
+				if len(outstanding) > 0xFFFF {
+					req.err <- errors.New("tag pool depleted")
+				} else if len(outstanding) == 0xFFFF {
+					if _, exists := outstanding[NOTAG]; !exists {
+						req.err <- errors.New("tag pool depleted")
+					}
+				}
+
+				exists := true
+				for selected = 0; exists && selected < NOTAG; selected++ {
```

Comment:
This does a linear search through the tag space, making tag allocation more expensive as there are more outstanding messages. If we keep track of `selected`, this can be avoided. Collisions can be managed with linear probing (odd, I know, but this is what is happening), where if a collision is found, we go to the next entry. 

Consider a slice. The maximum bound for memory usage per session for the tag, aside from hash overhead (fill factor `λ`, where `λ` < 1) and the value side of the hash, will be ~`(N(size(p)+size(tag))/λ`, which is to 640K on 64-bit with `λ = 1`, and larger with a more realistic fill factor. With a slice, the memory becomes bounded by `N*size(p)`, which is just 512KB on 64-bit. We will also benefit in look since the values are linearly arranged in memory, which is optimal with a linear probing strategy. The trade off here is that holes cost us, but holes exist only as long as outstanding messages still exist. The holes will be proportional in size to the age of the oldest outstanding message.